dvc/api.py

```diff
@@ -20,17 +21,22 @@ def get_url(path, repo=None, rev=None, remote=None):
     in a DVC repo. For Git repos, HEAD is used unless a rev argument is
     supplied. The default remote is tried unless a remote argument is supplied.
 
-    Raises UrlNotDvcRepoError if repo is not a DVC project.
+    Raises OutputNotFoundError if the file is not a dvc-tracked file.
 
     NOTE: This function does not check for the actual existence of the file or
     directory in the remote storage.
     """
     with _make_repo(repo, rev=rev) as _repo:
-        if not isinstance(_repo, Repo):
-            raise UrlNotDvcRepoError(_repo.url)  # pylint: disable=no-member
-        out = _repo.find_out_by_relpath(path)
-        remote_obj = _repo.cloud.get_remote(remote)
-        return str(remote_obj.tree.hash_to_path_info(out.checksum))
+        tree = RepoTree(_repo, fetch=True)
```

Comment:
```suggestion
        tree = DvcTree(_repo, fetch=True)
```

Reply:
`DvcTree` is mostly internal, and does not support subrepos.

Reply:
Okay, I think I get why you are suggesting that. As `DvcTree` does not support subrepos, we need to implement `RepoTree.get_dvctree(path)` or `RepoTree.get_repo(repo_path)`, as `get_hash` also hashes git-files.

Reply:
Or, `get_hash(dvc_only=True)`? Thoughts?

Reply:
Yeah, I suggest it because this is not meant to work with git files. Not a big fan of `dvc_only=True` though, but we could simply do:
```
if not tree.isdvc(path):
    raise ....
hash = tree.get_hash() 
```
:slightly_smiling_face:

Reply:
Theoretically we could've made hash part of the metadata, but that might make it complicated for files inside dirs, as you'll need to parse the dir_cache (as we do in _get_granular_checksum, which is not that bad)

Reply:
Btw, this will automatically close #3182  :slightly_smiling_face:

Reply:
Okay, did as you recommended, but `metadata`/`hash`/`RepoTree` all gets mixed. Feels like that thing could be a part of `RepoTree` on itself.